patterns/behavioural/command.md

```diff
@@ -0,0 +1,231 @@
+}
+```
+
+## Approach: Using function pointers
+
+We could follow another solution by
+creating each individual command as
+a different function and store function pointers
+to invoke these functions later at a different time.
+Since function pointers implement all three traits
+`Fn`, `FnMut`, and `FnOnce` we could as well pass and store
+closures instead of function pointers.
+
+```rust
+type FnPtr<'a> = fn() -> &'a str;
+struct Command<'a> {
+    execute: FnPtr<'a>,
+    rollback: FnPtr<'a>,
+}
+
+struct Schema<'a> {
+    commands: Vec<Box<Command<'a>>>,
+}
+
+impl<'a> Schema<'a> {
+    fn new() -> Self {
+        Self { commands: vec![] }
+    }
+    fn add_migration(&mut self, execute: FnPtr<'a>, rollback: FnPtr<'a>) {
+        self.commands.push(Box::new(Command { execute, rollback }));
+    }
+    fn execute(&self) -> Vec<&str> {
+        self.commands.iter().map(|cmd| (cmd.execute)()).collect()
+    }
+    fn rollback(&self) -> Vec<&str> {
+        self.commands
+            .iter()
+            .rev()
+            .map(|cmd| (cmd.rollback)())
+            .collect()
+    }
+}
+
+fn add_field() -> &'static str {
+    "add field"
+}
+
+fn remove_field() -> &'static str {
+    "remove field"
+}
+
+fn main() {
+    let mut schema = Schema::new();
+    schema.add_migration(|| "create table", || "drop table");
+    schema.add_migration(add_field, remove_field);
+    assert_eq!(vec!["create table", "add field"], schema.execute());
+    assert_eq!(vec!["remove field", "drop table"], schema.rollback());
+}
+```
+
+## Approach: Using `Fn` trait objects
+
+Finally, instead of defining a common command
+trait we simply store each command implementing
+the `Fn` trait separately in vectors.
+
+```rust
+type Migration<'a> = Box<dyn Fn() -> &'a str>;
+
+struct Schema<'a> {
+    executes: Vec<Migration<'a>>,
+    rollbacks: Vec<Migration<'a>>,
+}
+
+impl<'a> Schema<'a> {
+    fn new() -> Self {
+        Self {
+            executes: vec![],
+            rollbacks: vec![],
+        }
+    }
+    fn add_migration<E, R>(&mut self, execute: E, rollback: R)
+    where
+        E: Fn() -> &'a str + 'static,
+        R: Fn() -> &'a str + 'static,
+    {
+        self.executes.push(Box::new(execute));
+        self.rollbacks.push(Box::new(rollback));
+    }
+    fn execute(&self) -> Vec<&str> {
+        self.executes.iter().map(|cmd| cmd()).collect()
+    }
+    fn rollback(&self) -> Vec<&str> {
+        self.rollbacks.iter().rev().map(|cmd| cmd()).collect()
+    }
+}
+
+fn add_field() -> &'static str {
+    "add field"
+}
+
+fn remove_field() -> &'static str {
+    "remove field"
+}
+
+fn main() {
+    let mut schema = Schema::new();
+    schema.add_migration(|| "create table", || "drop table");
+    schema.add_migration(add_field, remove_field);
+    assert_eq!(vec!["create table", "add field"], schema.execute());
+    assert_eq!(vec!["remove field", "drop table"], schema.rollback());
+}
+```
+
+## Discussion
```

Comment:
A discussion when to use the third approach (`Using `Fn` trait objects`) would be nice as well.